[Time series] Add PatchTSMixer

ajati opened the pull request:
[PatchTSMixer](https://arxiv.org/pdf/2306.09364.pdf) ([KDD 2023](https://dl.acm.org/doi/abs/10.1145/3580305.3599533)) is a lightweight time-series modeling approach based on the MLP-Mixer architecture. In this HuggingFace implementation, we provide PatchTSMixer's capabilities to effortlessly facilitate lightweight mixing across patches, channels, and hidden features for effective multivariate time-series modeling. It also supports various attention mechanisms starting from simple gated attention to more complex self-attention blocks that can be customized accordingly. The model can be pretrained and subsequently used for various downstream tasks such as forecasting, classification and regression.

@kashif 

Done: ~~TODOs~~

- [x] Add generate method
- [x] Make pretrained dataset publicly available
- [x] Make pretrained weights publicly available

ajati committed:
patchtsmixer initial commit

ajati committed:
x,y->context_values,target_values, unittest addded

vijaye12ibm committed:
cleanup code

ajati committed:
minor

vijaye12ibm committed:
return hidden states

ajati committed:
model tests, partial integration tests

ajati committed:
ettm notebook temporary

ajati committed:
minor

ajati committed:
config mask bug fix, tests updated

ajati committed:
final ETT notebooks

vijaye12ibm committed:
add selfattn

vijaye12ibm committed:
init

ajati committed:
added docstrings

ajati committed:
PatchTSMixerForPretraining -> PatchTSMixerForMaskPretraining

ajati committed:
functionality tests added

ajati committed:
add start and input docstrings

vijaye12ibm committed:
docstring edits

vijaye12ibm committed:
testcase edits

vijaye12ibm committed:
minor changes

ajati committed:
docstring error fixed

ajati committed:
ran make fixup

ajati committed:
finalize integration tests and docs

ajati committed:
minor

ajati committed:
Merge branch 'main' into add-patchtsmixer

ajati committed:
cleaned gitignore

ajati committed:
added dataclass decorator, ran black formatter

ajati committed:
ran ruff

kashif committed:
Merge branch 'main' into add-patchtsmixer

kashif committed:
formatting

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26247). All of your documentation changes will be reflected on that endpoint.

kashif committed:
add slow decorator

kashif committed:
renamed in_Channel to input_size and default to 1

kashif committed:
shorten dataclass names

kashif committed:
use smaller model for testing

kashif committed:
moved the 3 heads to the modeling file

kashif committed:
use scalers instead of revin

kashif committed:
support forecast_channel_indices

kashif committed:
fix regression scaling

kashif committed:
undo reg. scaling

kashif committed:
removed unneeded classes

kashif committed:
forgot missing

kashif committed:
add more layers

kashif committed:
add copied positional_encoding

kashif committed:
use patchmask from patchtst

ajati committed:
removed dependency on layers directory

kashif committed:
formatting

kashif committed:
set seed

kashif committed:
Merge branch 'main' into add-patchtsmixer

kashif commented:
depends on #25927

ArthurZucker commented:
I am not sure I understand should I review this ? It's all red, while the other PR does not seem ready either (a lot of recent commits + red as well) ?

kashif committed:
removed unused imports

ajati committed:
fixed forward signature test

vijaye12ibm committed:
adding distributional head for PatchTSMixerForecasting

vijaye12ibm committed:
add generate to forecast

vijaye12ibm committed:
testcases for generate

vijaye12ibm committed:
add generate and distributional head for regression

vijaye12ibm committed:
raise Exception for negative values for neg binominal distribution

vijaye12ibm committed:
formatting changes

vijaye12ibm committed:
remove copied from patchtst and add TODO for test passing

vijaye12ibm committed:
make copies

vijaye12ibm committed:
doc edits

vijaye12ibm committed:
minor changes

vijaye12ibm committed:
format issues

vijaye12ibm committed:
minor changes

vijaye12ibm committed:
minor changes

namctin committed:
format docstring

namctin committed:
change some class names to PatchTSMixer + class name

Transpose to PatchTSMixerTranspose
GatedAttention to PatchTSMixerGatedAttention

namctin committed:
change NormLayer to PatchTSMixerNormLayer

namctin committed:
change MLP to PatchTSMixerMLP

namctin committed:
change PatchMixer to PatchMixerBlock, FeatureMixer to FeatureMixerBlock

namctin committed:
change ChannelFeatureMixer to ChannelFeatureMixerBlock

namctin committed:
change PatchMasking to PatchTSMixerMasking

namctin committed:
change Patchify to PatchTSMixerPatchify

namctin committed:
list to `list`

kashif committed:
fix docstrings

kashif committed:
formatting

namctin committed:
change bs to batch_size, edit forecast_masking

namctin committed:
Merge branch 'add-patchtsmixer' of https://github.com/namctin/transformers into add-patchtsmixer

namctin committed:
edit random_masking

namctin committed:
change variable name and update docstring in PatchTSMixerMasking

namctin committed:
change variable name and update docstring in InjectScalerStatistics4D

namctin committed:
update forward call in PatchTSMixerTranspose

namctin committed:
change variable name and update docstring in PatchTSMixerNormLayer

namctin committed:
change variable name and update docstring in PatchTSMixerMLP

namctin committed:
change variable name and update docstring in ChannelFeatureMixerBlock

vijaye12 committed:
formatting

vijaye12 committed:
formatting issues

vijaye12 committed:
docstring issue

kashif committed:
fixed observed_mask type in docstrings

kashif committed:
use FloatTensor type

vijaye12 committed:
formatting

ajati committed:
fix rescaling issue in forecasting, fixed integration tests

kashif committed:
add docstring from decorator

kashif committed:
fix docstring

NielsRogge commented on README.md:
```diff
@@ -433,6 +433,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[OpenLlama](https://huggingface.co/docs/transformers/model_doc/open-llama)** (from [s-JoL](https://huggingface.co/s-JoL)) released in [Open-Llama](https://github.com/s-JoL/Open-Llama).
 1. **[OPT](https://huggingface.co/docs/transformers/master/model_doc/opt)** (from Meta AI) released with the paper [OPT: Open Pre-trained Transformer Language Models](https://arxiv.org/abs/2205.01068) by Susan Zhang, Stephen Roller, Naman Goyal, Mikel Artetxe, Moya Chen, Shuohui Chen et al.
 1. **[OWL-ViT](https://huggingface.co/docs/transformers/model_doc/owlvit)** (from Google AI) released with the paper [Simple Open-Vocabulary Object Detection with Vision Transformers](https://arxiv.org/abs/2205.06230) by Matthias Minderer, Alexey Gritsenko, Austin Stone, Maxim Neumann, Dirk Weissenborn, Alexey Dosovitskiy, Aravindh Mahendran, Anurag Arnab, Mostafa Dehghani, Zhuoran Shen, Xiao Wan ...
+1. **[PatchTSMixer](model_doc/patchtsmixer)** (from IBM) released with the paper [TSMixer: Lightweight MLP-Mixer Model for Multivariate Time Series Forecasting](https://arxiv.org/abs/2306.09364) by Vijay Ekambaram, Arindam Jati, Nam Nguyen, Phanwadee Sinthong, Jayant Kalagnanam.
```
```suggestion
1. **[PatchTSMixer](https://huggingface.co/docs/transformers/main/model_doc/patchtsmixer)** (from IBM) released with the paper [TSMixer: Lightweight MLP-Mixer Model for Multivariate Time Series Forecasting](https://arxiv.org/abs/2306.09364) by Vijay Ekambaram, Arindam Jati, Nam Nguyen, Phanwadee Sinthong, Jayant Kalagnanam.
```
Make sure to run `make fix-copies` to update the other READMEs as well

kashif committed:
Update README.md

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,263 @@
+
+        self.head_dropout = head_dropout
+
+        self.num_patches = (max(seq_len, patch_len) - patch_len) // stride + 1
```
This can be calculated inside the model

NielsRogge commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,263 @@
+    based on the input data requirements. PatchTSMixer backbone can currently be extended for various downstream tasks
+    such as forecasting, classification and regression.
+
+    This is the configuration class to store the configuration of an [`PatchTSMixerModel`]. It is used to instantiate
+    an PatchTSMixer model according to the specified arguments, defining the model architecture. Instantiating a
```
```suggestion
    This is the configuration class to store the configuration of an [`PatchTSMixerModel`]. It is used to instantiate
    a PatchTSMixer model according to the specified arguments, defining the model architecture. Instantiating a
```

kashif committed:
Update src/transformers/models/patchtsmixer/configuration_patchtsmixer.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+
+class PatchTSMixerTranspose(nn.Module):
+    """
+    Parameters:
+    Transpose the tensor to the dimension defined in **dims**
+        dims (`list`): list of dimensions to be transposed contiguous (`bool`): if True, the transposed tensor is
+        contiguous
```
```suggestion
    Transpose the tensor to the dimension defined in **dims**
    
    Args:
        dims (`list`, *optional*, defaults to `False`):
               List of dimensions to be transposed contiguous. if True, the transposed tensor is contiguous.
```

kashif committed:
Update src/transformers/models/patchtsmixer/modeling_patchtsmixer.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            self.norm_attn = PatchTSMixerNormLayer(norm_mlp=norm_mlp, mode=mode, num_features=num_features)
+
+    def forward(self, x):
+        # x.shape == (batch_size, num_patches, num_features) if flatten
+        # x.shape == (batch_size, n_vars, num_patches, num_features) if common_channel
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            )
+            self.norm_attn = PatchTSMixerNormLayer(norm_mlp=norm_mlp, mode=mode, num_features=num_features)
+
+    def forward(self, x):
```
no one-letter variable names please

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/4a8f3a3e5b395bba2708a5b364452ab6051642ca

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        if self.gated_attn:
+            x = self.gab(x)
+
+        # if self.gated_attn:
+        #     attn_weight = self.attn_softmax(self.attn_layer(x))
+        #     x = x * attn_weight
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        elif self.mode in ["common_channel", "mix_channel"]:
+            x = x.transpose(2, 3)
+
+        # x.shape == (batch_size, num_patches, num_features) if flatten
+        # x.shape == (batch_size, n_vars, num_patches, num_features) if common_channel
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+    """FeatureMixerBlock
+
+    Args:
+        num_features (`int`, *optional*): Hidden feature size. Defaults to 16.
```
docstring doesn't follow the HF format here

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            self.gab = PatchTSMixerGatedAttention(in_size=num_features, out_size=num_features)
+
+    def forward(self, x):
+        # x.shape == (batch_size, num_patches, num_features) if flatten
+        # x.shape == (batch_size, n_vars, num_patches, num_features) if common_channel
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+
+        x = self.norm(x)
+
+        # if self.gated_attn:
+        #     x = self.gab(x)
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        if self.gated_attn:
+            self.gab = PatchTSMixerGatedAttention(in_size=num_features, out_size=num_features)
+
+    def forward(self, x):
```
no one-letter variable names please

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/4a8f3a3e5b395bba2708a5b364452ab6051642ca

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+class PatchTSMixerBackbone(nn.Module):
+    """PatchTSMixerBackbone
+
+    Args:
```
docstring doesn't follow the HF format

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        x = self.patch_mixer(x)
+        x = self.feature_mixer(x)
+
+        # x.shape == (batch_size, num_patches, num_features)
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            )
+
+    def forward(self, x):
+        # x.shape == (batch_size, num_patches, num_features)
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+                ffn=ffn,
+            )
+
+    def forward(self, x):
```
no one-letter variable names please

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/4a8f3a3e5b395bba2708a5b364452ab6051642ca

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            ]
+        )
+
+    def forward(self, x, output_hidden_states: Optional[bool] = False):
```
no one-letter variable names please

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/4a8f3a3e5b395bba2708a5b364452ab6051642ca

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+
+        all_hidden_states = []
+
+        logger.debug(x.shape)
```
can be removed

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/46360d694e1e7f4464b268251bc8b6132db00461

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        ffn = "mlp"
+        mixer_type = "base"
+
+        # if mode == "flatten":
+        #     logger.warn("Use mode = common_channel or mix_channel. mode=flatten is not preferred due to poor performance")
```
can be updated or removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        if use_pe:
+            self.W_pos = positional_encoding(pe, learn_pe, num_patches, num_features)
+
+    def forward(self, x, output_hidden_states: Optional[bool] = False):
```
no one-letter variable names please

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/4a8f3a3e5b395bba2708a5b364452ab6051642ca

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+
+        embedding, all_hidden_states = self.mlp_mixer_encoder(patches, output_hidden_states=output_hidden_states)
+
+        logger.debug(x.shape)
+        # embedding.shape == (batch_size, num_patches, num_features) if flatten
+        # embedding.shape == (batch_size, n_vars, num_patches, num_features) if common_channel
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        logger.debug(x.shape)
+
+        if self.use_pe:
+            patches = patches + self.W_pos
```
W_pos is not the best variable name, so does use_pe => let's use the same names as the other PR

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/49a485b35b85da165ec5b0f0ad95e8f500efcaec

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            x
+        )  # flatten: [bs x num_patch x num_features]   common_channel/mix_channel: [bs x n_vars x num_patch x num_features]
+
+        logger.debug(x.shape)
```
can be removed

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/46360d694e1e7f4464b268251bc8b6132db00461

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        # elif self.mode in ["common_channel", "mix_channel"]:
+        #     x = x.permute(0, 2, 1, 3)  # x: [bs x n_vars x num_patch x patch_len]
+
+        logger.debug(x.shape)
```
can be removed

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/46360d694e1e7f4464b268251bc8b6132db00461

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+    """Forecast Head
+
+    Args:
+        num_patches (`int`): Number of patches to segment
```
docstring doesn't follow the HF format, and you can prefix the class with PatchTST

NielsRogge replied:
Also, if the head is the same as the PatchTST PR, let's use copied from once the other PR is merged

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+
+            self.flatten = nn.Flatten(start_dim=1)
+
+    def forward(self, x, y=None):
```
no x, y please

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/4a8f3a3e5b395bba2708a5b364452ab6051642ca

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+        return forecast
+
+
+class LinearHead(nn.Module):
```
same here but I guess this can be copied from once the other PR gets merged

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+            module.gradient_checkpointing = value
+
+
+class PretrainHead(nn.Module):
```
same here, let's use copied from

vijaye12 replied:
done

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2361 @@
+
+
+# TODO: add copied from after PatchTST master merge
+def positional_encoding(pe, learn_pe, q_len, d_model):
```
same here, let's use copied from

patrickvonplaten replied:
Yes this function looks to be exactly the same as here
```suggestion
# Copied from transformers.patchtst.modeling_patchtst.positional_encoding
def positional_encoding(pe, learn_pe, q_len, d_model):
```

patrickvonplaten replied:
Ah ok there is already a comment!

vijaye12 replied:
corrected

NielsRogge commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,263 @@
+
+class PatchTSMixerConfig(PretrainedConfig):
+    r"""
+    PatchTSMixer is a lightweight time-series modelling backbone based on the MLP-Mixer architecture. Our [KDD 2023
```
```suggestion
    PatchTSMixer is a lightweight time-series modelling backbone based on the MLP-Mixer architecture. The [KDD 2023
```

NielsRogge commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,263 @@
+        forecast_channel_indices (`list`, *optional*):
+            List of channel indices to forecast. If None, forecast all channels. Target data is expected to have all
+            channels and we explitly filter the channels in prediction and target before loss computation.
+        n_classes (`int`, *optional*, defaults to 3):
+            Number of classes for a classification task.
+        n_targets (`int`, *optional*, defaults to 3):
+            Number of targets (dimensionality of the regressed variable) for a regression task.
```
see the PatchTST PR, one can use the default `num_labels` attribute for this

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+
+
+class PatchTSMixerGatedAttention(nn.Module):
+    """PatchTSMixerGatedAttention
```
no need to repeat the class name in the docstring

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+    """PatchTSMixerGatedAttention
+
+    Args:
+        in_size (`int`): input size
```
doesn't follow the HF format

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        if "batch" in self.norm_mlp.lower():
+            if self.mode in ["common_channel", "mix_channel"]:
+                # reshape the data
+                x_tmp = torch.reshape(
```
let's use a better name than x_tmp

NielsRogge commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,263 @@
+}
+
+
+class PatchTSMixerConfig(PretrainedConfig):
```
I'd double check the comments I made on the other PR as they apply to this config as well

vijaye12 replied:
corrected

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
(nit) we rather use nn.functional directly rather than using a one-letter import

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+
+PATCHTSMIXER_INPUTS_DOCSTRING = r"""
+    Parameters:
+        context_values (`torch.FloatTensor` of shape `(batch_size, seq_length, input_size)`):
```
aren't we using `past_values` for all of our time series models?

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/e1e8d2095f8ecb46657d5009a40e8a25232f2232

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+    
+    Args:
+        dims (`list`, *optional*, defaults to `False`):
+               List of dimensions to be transposed contiguous. if True, the transposed tensor is contiguous.
```
can be indented to the left

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        num_features=16,
+    ):
+        super().__init__()
+        self.norm_mlp = norm_mlp
```
if only layernorm is used, I'd remove the norm_mlp argument and just use nn.LayerNorm here

ajati replied:
BatchNorm is also used, as can be seen below the marked lines.

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+    """ChannelFeatureMixerBlock
+
+    Args:
+        num_features (`int`, *optional*): Hidden feature size. Defaults to 16.
```
docstring doesn't follow the HF format

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+
+
+class ChannelFeatureMixerBlock(nn.Module):
+    """ChannelFeatureMixerBlock
```
no need to repeat the class name here

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        return inputs
+
+
+class ChannelFeatureMixerBlock(nn.Module):
```
```suggestion
class PatchTSMixerChannelFeatureMixerBlock(nn.Module):
```

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        residual = inputs
+        inputs = self.norm(inputs)
+
+        # inputs.shape == (batch_size, num_features, num_patches, num_channels)
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+    """PatchMixerBlock
+
+    Args:
+        num_features (`int`, *optional*): Hidden feature size. Defaults to 16.
```
docstring doesn't follow the HF format

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        return out
+
+
+class FeatureMixerBlock(nn.Module):
+    """FeatureMixerBlock
```
no need to repeat class name in docstring, and can be prefixed by PatchTSMixer

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+
+    def __init__(
+        self,
+        num_patches: int,
```
typically our classes just take the `config` as input, rather than all individual args.

Not sure if possible here?

ajati replied:
Hi @NielsRogge , thanks for the comment. These are some classes that are used internally. The main exposed classes which are generally inherited from `PatchTSMixerPreTrainedModel`, take `config` as the input. Is this okay?

NielsRogge replied:
We usually also do this for internal classes as well

vijaye12 replied:
Sure @NielsRogge We enabled this change to all the classes that get initialized with the config params. We now pass it as config, instead of passing every param individually. 

@ajati - this requirement had a lot of code refactoring. Please review it carefully from your side too.

ajati replied:
Sure.

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        return x
+
+
+class PatchTSMixerBackbone(nn.Module):
```
```suggestion
class PatchTSMixerEncoder(nn.Module):
```
We typically use the term `Encoder` for this

ajati replied:
Hi @NielsRogge , thanks for the suggestion. We already have a `PatchTSMixerEncoder` which is used by the `PatchTSMixerModel` class. The `PatchTSMixerModel` is exposed. On the other hand, `PatchTSMixerBackbone` is used by the `PatchTSMixer` class which is not exposed (just internally used).

vijaye12 replied:
I think, the work backbone is confusing here. I renamed it to PatchTSMixerBlock

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+        """
+        if self.mode in ["common_channel", "mix_channel"]:
+            x = self.flatten(x)  # [batch_size x n_vars x num_patch * num_features]
+            # x = torch.reshape(
+            #     x, (x.shape[0], x.shape[1], x.shape[2] * x.shape[3])
+            # )  # [batch_size x n_vars x num_patch * num_features]
```
can be removed

NielsRogge commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+    loss: Optional[torch.FloatTensor] = None
+
+
+class PatchTSMixerForMaskPretraining(PatchTSMixerPreTrainedModel):
```
see other PR, let's use the same term as BertForPretraining
```suggestion
class PatchTSMixerForPretraining(PatchTSMixerPreTrainedModel):
```

NielsRogge reviewed (commented):
Main comments:

- updating docstrings to HF format
- use clear variable names everywhere
- use copied from wherever possible, especially after #25927 is merged
- do not use `context_values` if we already use the term `past_values` for all our time series models (keep in mind we can't change this afterwards)

kashif committed:
Update src/transformers/models/patchtsmixer/configuration_patchtsmixer.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

kashif committed:
Update src/transformers/models/patchtsmixer/modeling_patchtsmixer.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

kashif committed:
PatchTSMixerChannelFeatureMixerBlock

kashif committed:
formatting

kashif committed:
ForPretraining

kashif committed:
use num_labels instead of n_classes

kashif committed:
remove commented out code

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2362 @@
+
+
+class PatchTSMixerBackbone(nn.Module):
+    """PatchTSMixerBackbone
```
Same as https://github.com/huggingface/transformers/pull/26247#discussion_r1360296662

ajati committed:
docstring fixed

ajati committed:
nn.functional used instead of one letter F

ajati committed:
x_tmp renamed

ajati committed:
one letter variable x removed from forward calls

ajati committed:
one letter variable y removed

ajati committed:
remove commented code

vijaye12 committed:
rename patch_size, in_channels, PatchTSMixerBackbone

vijaye12 committed:
add config to heads

vijaye12 committed:
add config to heads tests

vijaye12 committed:
code reafactoring to use config instead of passing individual params

ajati committed:
Cdocstring fixes part 1

ajati committed:
docstring fixes part 2

ajati committed:
removed logger.debug

ajati committed:
context_values -> past_values

vijaye12 commented:
> Main comments:
> 
> * updating docstrings to HF format
> * use clear variable names everywhere
> * use copied from wherever possible, especially after [[time series] Add PatchTST #25927](https://github.com/huggingface/transformers/pull/25927) is merged
> * do not use `context_values` if we already use the term `past_values` for all our time series models (keep in mind we can't change this afterwards)

Hi @NielsRogge - We have completed all your suggested changes. Requesting your review.  Feel free to let us know if any issue is missed out. Only pending step is "copied from PatchTST" which we will bring it back as soon as PatchTST is merged.. CC: @kashif @ajati @namctin

patrickvonplaten commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,259 @@
+        use_pe (`bool`, *optional*, defaults to `False`):
+            Enable the use of positional embedding for the tiny self-attention layers. Works only when `self_attn` is
+            set to `True`.
+        pe (`str`, *optional*, defaults to `"zeros"`):
```
```suggestion
        positional_encoding (`str`, *optional*, defaults to `"zeros"`):
```
I'd prefer to write it out here, `"pe"` is not that known IMO.

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/49a485b35b85da165ec5b0f0ad95e8f500efcaec

patrickvonplaten commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,259 @@
+            across patches.
+        self_attn_heads (`int`, *optional*, defaults to 1):
+            Number of self-attention heads. Works only when `self_attn` is set to `True`.
+        use_pe (`bool`, *optional*, defaults to `False`):
```
```suggestion
        use_positional_encoding (`bool`, *optional*, defaults to `False`):
```

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/49a485b35b85da165ec5b0f0ad95e8f500efcaec

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+import torch
+import torch.nn as nn
+import torch.utils.checkpoint
+from torch.nn.modules.activation import MultiheadAttention
```
I'm ok with using PyTorch's `MultiheadAttention` layer. It'll just make very difficult to add FlashAttentionV2 with it afterward. @ArthurZucker wdyt?

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+
+    def __init__(self, *dims, contiguous=False):
+        super().__init__()
+        self.dims, self.contiguous = dims, contiguous
```
```suggestion
        self.dims = dims
        self.contiguous = contiguous
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+
+    """
+
+    def __init__(
+        self,
+        config: PatchTSMixerConfig,
+    ):
```
```suggestion
    def __init__(self, config: PatchTSMixerConfig):
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        mode = config.mode
+        norm_mlp = config.norm_mlp
+
+        self.norm_mlp = norm_mlp
+        self.mode = mode
```
```suggestion
```
not need I think

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+
+    """
+
+    def __init__(
+        self,
+        config: PatchTSMixerConfig,
+    ):
```
```suggestion
    def __init__(self, config: PatchTSMixerConfig):
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            Configuration.
+    """
+
+    def __init__(
+        self,
+        config: PatchTSMixerConfig,
+    ):
```
```suggestion
    def __init__(self, config: PatchTSMixerConfig):
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            ]
+        )
+
+    def forward(self, data, output_hidden_states: Optional[bool] = False):
```
```suggestion
    def forward(self, data, output_hidden_states: bool = False):
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            Configuration.
+    """
+
+    # @get_class_params
+    def __init__(
+        self,
+        config: PatchTSMixerConfig,
+    ):
```
```suggestion
    def __init__(self, config: PatchTSMixerConfig):
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        if use_pe:
+            self.W_pos = positional_encoding(pe, learn_pe, num_patches, num_features)
+
+    def forward(self, input_ts, output_hidden_states: Optional[bool] = False):
```
```suggestion
    def forward(self, input_ts, output_hidden_states: bool = False):
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+
+            self.flatten = nn.Flatten(start_dim=1)
+
+    def forward(self, hidden_features, target=None):
```
```suggestion
    def forward(self, hidden_features):
```
This tensors doesn't seem to be used anywhere

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/9112cbf3979d1aa8528882039ffd5ebad00e1259

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+
+        self.dropout = nn.Dropout(head_dropout)
+
+    def forward(self, hidden_features, target=None):
```
```suggestion
    def forward(self, hidden_features:
```

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/9112cbf3979d1aa8528882039ffd5ebad00e1259

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            hidden_features (`torch.Tensor` of shape `(batch_size x num_patch x num_features)` in `flatten` mode
+                or `(batch_size x n_vars x num_patch x num_features)` in `common_channel`/`mix_channel` mode.): Input
+                hidden features.
+            target (`torch.Tensor`, *optional*):
+                Target tensor.
+
```
```suggestion
            target (`torch.Tensor`, *optional*):
                Target tensor.
```suggestion

```
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            hidden_features (`torch.Tensor` of shape `(batch_size x num_patch x num_features)` in `flatten` mode
+                or `(batch_size x n_vars x num_patch x num_features)` in `common_channel`/`mix_channel` mode.): Input
+                hidden features.
+            target (`torch.Tensor`, *optional*):
+                Target tensor.
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            hidden_features (`torch.Tensor` of shape `(batch_size x num_patch x num_features)` in `flatten` mode
+                or `(batch_size x n_vars x num_patch x num_features)` in `common_channel`/`mix_channel` mode.): Input
+                hidden features.
+            target (`torch.Tensor`, *optional*):
+                Target tensor.
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+                nn.Linear(num_features, patch_len * input_size),
+            )
+
+    def forward(self, hidden_features, target=None):
```
```suggestion
    def forward(self, hidden_features):
```

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/9112cbf3979d1aa8528882039ffd5ebad00e1259

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            hidden_features (`torch.Tensor` of shape `(batch_size x num_patch x num_features)` in `flatten` mode
+                or `(batch_size x n_vars x num_patch x num_features)` in `common_channel`/`mix_channel` mode.): Input
+                hidden features.
+            target (`torch.Tensor`, *optional*):
+                Target tensor.
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            return hidden_features
+        elif self.mode in ["common_channel", "mix_channel"]:
+            forecast = self.base_pt_block(hidden_features)  # [batch_size x n_vars x num_patch x patch_len]
+            return forecast
```
What happens if self.mode is neither of the three? Can we raise here ? 
```suggestion
            return forecast
        else:
            raise ValueError(f"self.mode has to be one of ['flatten', 'common_channel', 'mix_channel'], but is {self.mode}.")
```

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/9112cbf3979d1aa8528882039ffd5ebad00e1259

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        self.minimum_scale = minimum_scale
+        self.default_scale = default_scale
+
+    @torch.no_grad()
```
same comments as in other PR

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+    def __init__(self, config: PatchTSMixerConfig):
+        super().__init__(config)
+
+        self.encoder = PatchTSMixer(
```
```suggestion
        self.encoder = PatchTSMixer(config=config)
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        super().__init__(config)
+
+        self.encoder = PatchTSMixer(
+            config=config,
+        )
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        self,
+        past_values: torch.Tensor,
+        output_hidden_states: Optional[bool] = False,
+        return_dict: Optional[bool] = None,
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        # past_values: [batch_size  x n_vars x num_patches x patch_len]
+        # return: [batch_size x n_vars x num_patches x num_features]
+        last_hidden_state, hidden_states = self.encoder(past_values, output_hidden_states=output_hidden_states)
+        return PatchTSMixerEncoderOutput(last_hidden_state=last_hidden_state, hidden_states=hidden_states)
```
Can we add a `if return_dict: ...` else statement here?

ajati replied:
Hi @patrickvonplaten , we added `return_dict` to all the exposed classes. For the internal classes, we did not add. Is that okay?

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+
+        encoder_output = self.encoder(enc_input, output_hidden_states=output_hidden_states)
+
+        return PatchTSMixerModelOutput(
```
Can we add a `if return_dict: ...` else statement here?

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/3daac8886a75ab9c285c172c5bd5e6073973c4aa

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        if self.masked_loss is True and loss_val is not None:
+            loss_val = (loss_val.mean(dim=-1) * model_output.mask).sum() / (model_output.mask.sum() + 1e-10)
+
+        return PatchTSMixerForMaskPreTrainingOutput(
```
Can we add a `if return_dict: ...` else statement here?

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            self.loss = nn.MSELoss(reduction="mean")
+            self.distribution_output = None
+        else:
+            self.loss = nll
```
Can we move the loss functions to the forward method?

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/c36e6fbb90a78c87a0be18fc8d30d0e47dd77f7c

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+            loc = model_output.loc
+            scale = model_output.scale
+
+        return PatchTSMixerForForecastOutput(
```
Should we add a `return_dict` here?

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        else:
+            loss_val = None
+
+        return PatchTSMixerForClassificationOutput(
```
should we set a return_dict here?

patrickvonplaten commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        else:
+            loss_val = None
+
+        return PatchTSMixerForRegressionOutput(
```
return dict?

patrickvonplaten commented on tests/models/patchtsmixer/test_modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,973 @@
+    def test_resize_tokens_embeddings(self):
+        pass
+
+    def test_model_outputs_equivalence(self):
```
This test should pass IMO and should not be skipped? Being able to return tuples is VERY useful for `torch.compile` etc...

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/3daac8886a75ab9c285c172c5bd5e6073973c4aa

patrickvonplaten reviewed (commented):
Very nice PR!

Would be great if we can clean some of the `return_dict`, unused variables and add Copied from as soon as is #25927 merged.

ajati committed:
formatting changes

ajati committed:
pe -> positional_encoding

ajati committed:
removed unused target variable

ajati committed:
self.mode logic fixed

ajati committed:
formatting change

namctin committed:
edit docstring and var name

namctin committed:
change n_targets to num_targets

namctin committed:
rename input_size to num_input_channels

namctin committed:
add head names with prefix PatchTSMixer

namctin committed:
edit docstring in PatchTSMixerForRegression

ajati committed:
fix var name change in testcases

kashif committed:
add PatchTSMixerAttention

ajati committed:
return dict for all exposed classes, test cases added

ajati committed:
Merge branch 'add-patchtsmixer' of github.com:namctin/transformers into add-patchtsmixer

kashif committed:
format

vijaye12 committed:
move loss function to forward call

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2190 @@
+        self.head = LinearHead(
+            config=config,
+        )
+        self.loss = torch.nn.CrossEntropyLoss()
```
https://github.com/huggingface/transformers/pull/26247/commits/c36e6fbb90a78c87a0be18fc8d30d0e47dd77f7c

ajati committed:
Merge branch 'main' into add-patchtsmixer

ajati committed:
make style

vijaye12 committed:
adding return dict/tuple

ajati committed:
make repo-consistency

vijaye12 commented:
Hi @patrickvonplaten @NielsRogge - We have completed most of the requested code changes. Requesting your review and approval if things are in place. 

All test-cases are passing now. 

Copied from PatchTST is still pending and will be done as soon as the PatchTST branch is merged.

CC: @kashif @namctin @ajati

ArthurZucker commented:
I'll review 😉

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+            self.post_init()
+
+    @add_start_docstrings_to_model_forward(PATCHTSMIXER_INPUTS_DOCSTRING)
+    # @replace_return_docstrings(output_type=PatchTSMixerForRegressionOutput, config_class=_CONFIG_FOR_DOC)
```
to remove?

ajati replied:
https://github.com/huggingface/transformers/pull/26247/commits/ec6ea20945e4fa817f627c49844882d17084ecc8

ArthurZucker reviewed (commented):
I don't see a conversion script so I suppose the idea was to match 1-1 the original structure but let's avoid this if possible. 
The code would be a lot cleaner if we have 1 class for each mode mixing. Keep it simple and isolate what can be done in a single class, then use a mapping to get the properclass. Otherwise there is too much granularity, and too many if else controlled by config arguments. It's hard to follow and not really transformers like. 
It's fine when it's one layer but here it's for all forward all layers. `PatchTSMixerLinearHead` Is fine with me but there are ways to make it a bit readable 

Moreover depending on the mode it's most often just the input / output shape that changes. So for this we should control the linear with this arg see for example this function 
```python 
        if self.mode in ["common_channel", "mix_channel"]:
            self.base_pt_block = nn.Sequential(
                nn.Dropout(head_dropout),
                nn.Linear(num_features, patch_len),
            )
        else:
            self.base_pt_block = nn.Sequential(
                nn.Dropout(head_dropout),
                nn.Linear(num_features, patch_len * num_input_channels),
            )
```
Sorry for the time sensitivity and the late review! 
Open to discussions if these models are super good, would be a pity not to make them into a reference implementation!

vijaye12 committed:
remove flatten mode

vijaye12 commented:
> I don't see a conversion script so I suppose the idea was to match 1-1 the original structure but let's avoid this if possible. The code would be a lot cleaner if we have 1 class for each mode mixing. Keep it simple and isolate what can be done in a single class, then use a mapping to get the properclass. Otherwise there is too much granularity, and too many if else controlled by config arguments. It's hard to follow and not really transformers like. It's fine when it's one layer but here it's for all forward all layers. `PatchTSMixerLinearHead` Is fine with me but there are ways to make it a bit readable
> 
> Moreover depending on the mode it's most often just the input / output shape that changes. So for this we should control the linear with this arg see for example this function
> 
> ```python
>         if self.mode in ["common_channel", "mix_channel"]:
>             self.base_pt_block = nn.Sequential(
>                 nn.Dropout(head_dropout),
>                 nn.Linear(num_features, patch_len),
>             )
>         else:
>             self.base_pt_block = nn.Sequential(
>                 nn.Dropout(head_dropout),
>                 nn.Linear(num_features, patch_len * num_input_channels),
>             )
> ```
> 
> Sorry for the time sensitivity and the late review! Open to discussions if these models are super good, would be a pity not to make them into a reference implementation!

Yes. We removed "flatten" mode which is just a baseline and not used mostly. Now, we do not have else-if block in every forward call.

vijaye12 committed:
code refactoring

vijaye12 commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,259 @@
+    documentation from [`PretrainedConfig`] for more information.
+
+    Args:
+        seq_len (`int`, *optional*, defaults to 32):
+            The context/history length for the input sequence.
```
seq_len refers to the actual context length and not the max. Also, it's not related to position embedding.

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        dropout = config.dropout
+        mode = config.mode
+        gated_attn = config.gated_attn
+        norm_mlp = config.norm_mlp
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+
+        self.gated_attn = gated_attn
+        if gated_attn:
+            self.gab = PatchTSMixerGatedAttention(in_size=num_input_channels, out_size=num_input_channels)
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        self_attn_heads = config.self_attn_heads
+        norm_mlp = config.norm_mlp
+        self.norm_mlp = config.norm_mlp
+        self.mode = config.mode
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        self.norm_mlp = config.norm_mlp
+        self.mode = config.mode
+
+        self.norm = PatchTSMixerNormLayer(norm_mlp=norm_mlp, mode=mode, num_features=num_features)
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+
+        self.gated_attn = gated_attn
+        if gated_attn:
+            self.gab = PatchTSMixerGatedAttention(in_size=num_patches, out_size=num_patches)
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        Returns:
+            `torch.Tensor`: Transformed tensor.
+        """
+        residual = data
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        self.gated_attn = gated_attn
+
+        if self.gated_attn:
+            self.gab = PatchTSMixerGatedAttention(in_size=num_features, out_size=num_features)
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+            self.patcher = nn.Linear(num_input_channels * patch_len, num_features)
+
+        elif mode in ["common_channel", "mix_channel"]:
+            self.patcher = nn.Linear(patch_len, num_features)
```
resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        forecast_len = config.forecast_len
+        head_dropout = config.head_dropout
+        mode = config.mode
+        forecast_channel_indices = config.forecast_channel_indices
```
resolved

vijaye12 committed:
rename data

vijaye12 committed:
remove PatchTSMixer and keep only PatchTSMixerEncoder

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+        mask_input (`bool`, *optional*, defaults to `False`):
+            If True, Masking will be enabled. False otherwise.
```
We do masking, only when PatchTSMixerModel is called from `PatchTSMixerForPretraining` class. When PatchTSMixerModel is called from other classes (like forecasting, regression or classification), masking is not enabled. This flag `mask_input` is set to just detect the class from where PatchTSMixerModel  is being called to enable/disable mask.

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+                    distribution_output.get_parameter_projection(num_patches * num_features),
+                )
+
+            self.flatten = nn.Flatten(start_dim=1)
```
Hi @ArthurZucker - can we replace nn.sequential with nn.moduleList Is that acceptable?

ajati replied:
All `nn.Sequential` have been removed. https://github.com/huggingface/transformers/pull/26247/commits/fe5babb195ee97811e340f9d97fde11e3a7c0891

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+
+
+class PatchTSMixerEncoder(PatchTSMixerPreTrainedModel):
```
resolved.

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+    def __init__(self, config: PatchTSMixerConfig, mask_input: bool = False):
+        super().__init__(config)
+
+        self.encoder = PatchTSMixerEncoder(config)
```
Removed PatchTSMixer and using only PatchTSMixerEncoder. Resolved

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if self.config.masked_loss is True:
+            loss = torch.nn.MSELoss(reduction="none")
+        else:
+            loss = torch.nn.MSELoss(reduction="mean")
```
Hi @ArthurZucker - the loss was earlier defined in init only. We had a comment from @patrickvonplaten to move it to the forward call. What is the preferred approach? Can we have it in init or forward?

ArthurZucker replied:
my bad! Seems like we do this in the forward 🤗 sorry for the mistake

vijaye12 replied:
no problem..

ajati committed:
docstring fixes

ajati committed:
removed unused code

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        else:
+            if self.distribution_output:
+                # if self.config.distribution_output == "negative_binomial" and torch.any(target_values < 0):
+                #     raise Exception("target_values cannot be negative for negative_binomial distribution.")
```
https://github.com/huggingface/transformers/pull/26247/commits/7d7e9f43dd06bfa166d2c9029a8641949d1fb2f9

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        if self.config.forecast_channel_indices is not None:
+            if self.distribution_output:
+                # if self.config.distribution_output == "negative_binomial" and torch.any(target_values < 0):
+                #     raise Exception("target_values cannot be negative for negative_binomial distribution.")
```
https://github.com/huggingface/transformers/pull/26247/commits/7d7e9f43dd06bfa166d2c9029a8641949d1fb2f9

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )  # model_output: [batch_size x nvars x num_patch x num_features]
+        if isinstance(model_output, tuple):
+            model_output = PatchTSMixerModelOutput(*model_output)
```
Yes, we did this to make the code simpler. We access many returned variables, so using indices is difficult.

ajati committed:
format

ajati committed:
format

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        if config.scaling in ["std", "mean", True]:
+            if config.mode == "flatten":
+                raise ValueError("Scaling is not supported for classification task when mode == flatten")
+            self.inject_scale = InjectScalerStatistics4D(
```
Hi @ArthurZucker , this class is being used in multiple places. Hence, we wrapped it as a reusable class.

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+    sequences: torch.FloatTensor = None
+
+
+# Copied from transformers.models.time_series_transformer.modeling_time_series_transformer.nll
```
Since this approach was followed in other HF time series transformers, we followed the same.

ArthurZucker replied:
got it thanks!

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        new_sequence_length = patch_length + stride * (self.num_patches - 1)
+        self.s_begin = sequence_length - new_sequence_length
+
+    def forward(self, past_values: torch.Tensor):
```
This is copied from `PatchTST`, we will resolve it in that branch.

vijaye12 replied:
corrected

ajati committed:
remove contiguous and formatting changes

ajati committed:
remove model description from config

ajati committed:
replace asserts with ValueError

ajati commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+
+        self.mode = mode
+
+        self.norm = PatchTSMixerNormLayer(norm_mlp=norm_mlp, mode=mode, num_features=num_features)
```
Hi Arthur, we have removed the `nn.Sequential` from this. However, we are doing some explicit shape transformations, hence, having this class is helpful.

vijaye12 replied:
yes. config is used now. @ajati we can tag this as resolved.

ajati committed:
remove nn.Sequential from PatchTSMixerNormLayer

ajati committed:
replace if-else with map

ajati committed:
remove all nn.Sequential

ajati committed:
format

namctin committed:
Merge branch 'huggingface:main' into add-patchtsmixer

ajati committed:
formatting

ajati committed:
Merge branch 'add-patchtsmixer' of github.com:namctin/transformers into add-patchtsmixer

ajati committed:
fix gradient_checkpointing error after merge, and formatting

ajati committed:
make fix-copies

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+            return inputs.transpose(*self.dims)
+
+
+class PatchTSMixerNormLayer(nn.Module):
```
Hi Arthur, we have removed the nn.Sequential from this. However, we are doing some explicit shape transformations which gets repeated in many places without this class, hence, having this class could be helpful.

ArthurZucker replied:
got it 😉 thanks

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+                (
+                    batch_size,
+                    self.num_patches,
+                    self.num_input_channels * self.patch_len,
```
Since users can make mistakes in calculating the number of patches, we calculate it in the init rather than giving the control to the end user.

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+        Returns:
+
+        """
+        if self.config.loss == "mse":
+            loss = nn.MSELoss(reduction="mean")
+        elif self.config.loss == "nll":
+            loss = nll
+        else:
+            raise ValueError("Invalid loss function: Allowed values: mse and nll")
+
```
we had a request earlier to move the loss functions forward. Hence we moved it from init to forward. However, let us know if it has to be moved to init.

ArthurZucker replied:
Ok my bad here 😉

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+
+        """
+
+        loss = torch.nn.CrossEntropyLoss()
```
we had a request earlier to move the loss functions forward. Hence we moved it from init to forward. However, let us know if it has to be moved to init.

ArthurZucker replied:
My bad 🤗 Sinse target values are optional let's only declare it if we are gonna compute it

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2447 @@
+
+        """
+
+        if self.config.loss == "mse":
+            loss = nn.MSELoss(reduction="mean")
+        elif self.config.loss == "nll":
+            loss = nll
+        else:
+            raise ValueError("Invalid loss function: Allowed values: mse and nll")
```
we had a request earlier to move the loss functions forward. Hence we moved it from init to forward. However, let us know if it has to be moved to init.

vijaye12 commented on src/transformers/models/patchtsmixer/configuration_patchtsmixer.py:
```diff
@@ -0,0 +1,259 @@
+        forecast_channel_indices (`list`, *optional*):
+            List of channel indices to forecast. If None, forecast all channels. Target data is expected to have all
+            channels and we explitly filter the channels in prediction and target before loss computation.
+        num_targets (`int`, *optional*, defaults to 3):
```
resolved

vijaye12 replied:
Hi @ArthurZucker - We tried segmenting the docstring in to various section by adding

`> Section name` as pointed in the previous example. But, when make fix-copies is getting called, these lines get automatically removed.. We tried few other seperators like ##### but even they are getting removed automatically after fix-copies.

Any suggestions to correct it?

vijaye12 commented:
Hi @ArthurZucker - We have addressed most of your comments. 

TODO: CopiedFrom from PatchTST pending after its merge.

CC: @kashif @ajati @namctin

vijaye12 commented on docs/source/en/model_doc/patchtsmixer.md:
```diff
@@ -0,0 +1,69 @@
+
+*Transformers have gained popularity in time series forecasting for their ability to capture long-sequence interactions. However, their high memory and computing requirements pose a critical bottleneck for long-term forecasting. To address this, we propose TSMixer, a lightweight neural architecture exclusively composed of multi-layer perceptron (MLP) modules. TSMixer is designed for multivariate  ...
+
+
```
yes. we have a set of demo notebooks which we will point from here. Also, we are working on a  blog which will also be added here.

ajati replied:
added

vijaye12 commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -1122,6 +1123,18 @@
 
+MODEL_FOR_TIME_SERIES_CLASSIFICATION_MAPPING_NAMES = OrderedDict(
+    [
+        ("patchtsmixer", "PatchTSMixerForClassification"),
```
@ajati @kashif  - I think the query here is to replace `PatchTSMixerForClassification` to `PatchTSMixerForTimeSeriesClassification` Your thoughts?

ajati replied:
Done.

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+            )  # inputs_reshaped: [batch_size*num_channels, num_patches, num_features]
+            # put back data to the original shape
+
+            inputs = torch.reshape(
+                inputs_reshaped,
+                (
+                    inputs.shape[0],
+                    inputs.shape[1],
+                    inputs.shape[2],
+                    inputs.shape[3],
+                ),
+            )
```
```suggestion
            inputs = torch.reshape(inputs_reshaped,inputs.shape)
```

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+                ),
+            )  # inputs_reshaped: [batch_size*num_channels, num_patches, num_features]
+
+            inputs_reshaped = self.norm(
+                inputs_reshaped
+            )  # inputs_reshaped: [batch_size*num_channels, num_patches, num_features]
```
```suggestion
            # inputs_reshaped: [batch_size*num_channels, num_patches, num_features]
            inputs_reshaped = self.norm(inputs_reshaped)  
```

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+
+        if self.config.self_attn:
+            hidden_state_reshaped = hidden_state
+            hidden_state_reshaped = torch.reshape(
+                hidden_state,
+                (hidden_state.shape[0] * hidden_state.shape[1], hidden_state.shape[2], hidden_state.shape[3]),
+            )
+            #  (batch_size, n_vars, num_patches, num_features)
+
+            x_attn, _, _ = self.self_attn_layer(hidden_state_reshaped, output_attentions=False)
+
+            x_attn = torch.reshape(
+                x_attn, (hidden_state.shape[0], hidden_state.shape[1], hidden_state.shape[2], hidden_state.shape[3])
+            )
+            #  (batch_size, n_vars, num_patches, num_features) if common_channel
```
given that hidden_state.shape[i] is re-used, let's do something like 
```python
batch_size, n_vars, num_patches, num_features = hidden_state.shape
```
and then just re-use them

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+        super().__init__()
+
+        self.norm = PatchTSMixerNormLayer(config)
+        self.config = config
```
self.config is not usefu let's just store `self.use_gate_attention = self.config.gated_attn`

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+
+
+class PatchTSMixerMLP(nn.Module):
+    def __init__(self, in_features, out_features, expansion_factor, dropout, last_dropout=True):
```
expansion factor and dropout are always taken from the config

ArthurZucker replied:
and last dropout does not seem to be passed to this module no?

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+        self.patch_mixer = PatchMixerBlock(config=config)
+        self.feature_mixer = FeatureMixerBlock(config=config)
+
+        self.config = config
```
let's just store `slef.mode`

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+    ):
+        super().__init__()
+
+        self.config = config
```
we don't need to store everything!

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+
+    def __init__(
+        self,
+        mask_type: str = "random",
+        mask_ratio=0.5,
+        mask_patches: list = [2, 3],
+        mask_patch_ratios: list = [1, 1],
+        channel_consistent_masking: bool = False,
+        unmasked_channel_indices: list = None,
```
same here for whatever can be taken from the config

vijaye12 replied:
corrected

ArthurZucker commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+        distribution = self.distribution_output.distribution(outputs.prediction_logits)
+
+        # get samples
+        samples = [
+            distribution.sample() for _ in range(num_parallel_samples)
+        ]  # samples: list of [batch_size x num_targets]
```
```suggestion
        # samples: list of [batch_size x num_targets]
        samples = [distribution.sample() for _ in range(num_parallel_samples)]  
```

vijaye12 replied:
corrected

ArthurZucker commented on tests/models/patchtsmixer/test_modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,1076 @@
+
+                                self.check_module(task="forecast", params=params)
+
+    def test_classification(self):
+        for mode in ["common_channel", "mix_channel"]:
+            for self_attn in [True, False]:
+                for scaling in [True, False, "mean", "std"]:
+                    for gated_attn in [True, False]:
+                        for head_aggregation in ["max_pool", "avg_pool"]:
+                            params = self.__class__.params.copy()
+                            params.update(
+                                mode=mode,
+                                self_attn=self_attn,
+                                scaling=scaling,
+                                head_aggregation=head_aggregation,
+                                gated_attn=gated_attn,
+                            )
+                            # print(mode,self_attn,revin,gated_attn,head_aggregation)
+
+                            self.check_module(task="classification", params=params)
+
+    def test_regression(self):
+        for mode in ["common_channel", "mix_channel"]:
+            for self_attn in [True, False]:
+                for scaling in [True, False, "mean", "std"]:
+                    for gated_attn in [True, False]:
+                        for head_aggregation in ["max_pool", "avg_pool"]:
+                            for loss in ["mse", "nll"]:
+                                params = self.__class__.params.copy()
```
thing we should use parametrized for this

vijaye12 replied:
corrected

ArthurZucker commented on tests/models/patchtsmixer/test_modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,1076 @@
+
+                                    self.check_module(task="pretrain", params=params)
+
+        # for mode in ["flatten","common_channel","mix_channel"]:
+        #     for task in ["forecast","classification","regression","pretrain"]:
+        #         for self_attn in [True,False]:
+        #             for head_aggregation in ["max_pool","avg_pool"]:
+        #                 for mask_type in ["random","forecast"]:
+        #                     for masked_loss in [True, False]:
+        #                         for channel_consistent_masking in [True, False]:
+        #                             for revin in [True, False]:
+        #                                 for forecast_channel_indices in [None, [0,2]]:
```
to remove?

vijaye12 replied:
corrected

ArthurZucker commented on tests/models/patchtsmixer/test_modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,1076 @@
+        )
+        self.forecast_full_module(params)
+
+    # def test_forecast_full(self):
```
to remove?

vijaye12 replied:
corrected

ArthurZucker commented on tests/models/patchtsmixer/test_modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,1076 @@
+        )
+        self.assertEqual(output.last_hidden_state.shape, self.__class__.enc_output.shape)
+        self.assertEqual(output.loss.item() < 100, True)
+        # print("loss shape", output.loss, output.loss.shape)
```
to remove?

vijaye12 replied:
corrected

ArthurZucker commented on tests/models/patchtsmixer/test_modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,1076 @@
+        self.assertEqual(output.last_hidden_state.shape, self.__class__.enc_output.shape)
+        self.assertEqual(output.loss.item() < 100, True)
+
+        # print("loss shape", output.loss, output.loss.shape)
```
to remove?

vijaye12 replied:
corrected

ArthurZucker reviewed (commented):
Looks a lot better thanks all for bearing with me 🤗 
In principle looks good to me, left a few nits and will have to wait for other PR to be merged. 
I'll let @amyeroberts handle the last review and merge as I'm off for next week! 
Good work!

vijaye12 commented:
> Looks a lot better thanks all for bearing with me 🤗 In principle looks good to me, left a few nits and will have to wait for other PR to be merged. I'll let @amyeroberts handle the last review and merge as I'm off for next week! Good work!

Sure. Thank you @ArthurZucker for your comments. We will resolve the newer comments soon and will wait for the other PR to get merged.

kashif committed:
Merge branch 'main' into add-patchtsmixer

kashif committed:
remove comments

kashif committed:
reshape

kashif committed:
doesnt support gradient checkpointing

vijaye12ibm committed:
corect Patchify

vijaye12ibm committed:
masking updates

vijaye12ibm committed:
batchnorm copy from

vijaye12ibm committed:
format checks

vijaye12ibm committed:
scaler edits

vijaye12 committed:
remove comments

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+        x = x.unfold(
+            dimension=-2, size=self.patch_length, step=self.stride
+        )  # x: [batch_size x num_patches x num_input_channels x patch_length]
+        x = x.transpose(-2, -3).contiguous()  # x: [batch_size x num_input_channels x num_patches x patch_length]
```
corrected

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+    Args:
+    A class to patchify the time series sequence into different patches
+        sequence_length (`int`, *required*): Input sequence length. patch_length (`int`, *required*): Patch length.
+        stride (`int`, *required*): Stride between patches.
```
corrected

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+
+    mask = mask.unsqueeze(-1).repeat(
+        1, 1, 1, num_features
+    )  # mask: [batch_size x num_channels x num_patch x patch_len]
```
corrected

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+# TODO: add copied from after PatchTST master merge
+def positional_encoding(position_embedding_type, learned, q_len, d_model):
+    # Positional encoding
+    if position_embedding_type is None:
```
yes. we can keep these as different options, if that is accepted.

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+        config (`PatchTSMixerConfig`, *required*): Configuration.
+    """
+
+    def __init__(
+        self,
+        config: PatchTSMixerConfig,
+        distribution_output=None,
+    ):
```
corrected

vijaye12 committed:
format changes

vijaye12 committed:
remove self.config

vijaye12 committed:
correct class PatchTSMixerMLP(nn.Module):

vijaye12 committed:
makr fix

vijaye12 committed:
doc updates

kashif committed:
Merge branch 'main' into add-patchtsmixer

kashif committed:
fix-copies

vijaye12 committed:
scaler class correction

vijaye12 committed:
doc edits

vijaye12 committed:
scaler edits

kashif committed:
update readme with links

vijaye12 committed:
injectstatistics add

vijaye12 committed:
Merge branch 'add-patchtsmixer' of github.com:namctin/transformers into add-patchtsmixer

kashif committed:
Merge branch 'main' into add-patchtsmixer

kashif committed:
fix-copies

namctin committed:
add norm_eps option to LayerNorm

vijaye12 committed:
format changes

vijaye12 committed:
fix copies

vijaye12 committed:
correct make copies

vijaye12 commented on src/transformers/models/patchtsmixer/modeling_patchtsmixer.py:
```diff
@@ -0,0 +1,2265 @@
+    ):
+        super().__init__()
+
+        self.config = config
```
corrected

vijaye12 committed:
use parametrize

vijaye12 commented:
> Main comments:
> 
> * updating docstrings to HF format
> * use clear variable names everywhere
> * use copied from wherever possible, especially after [[time series] Add PatchTST #25927](https://github.com/huggingface/transformers/pull/25927) is merged
> * do not use `context_values` if we already use the term `past_values` for all our time series models (keep in mind we can't change this afterwards)

All corrections completed

kashif committed:
fix doc string

kashif committed:
add docs to toctree

kashif committed:
make style

vijaye12 committed:
doc segmenting

vijaye12 committed:
docstring edit

vijaye12 commented:
Hi @ArthurZucker  @NielsRogge @patrickvonplaten -We have resolved most of the feedback given in the last round. PR is ready for your review.
CC: @kashif

kashif commented:
@amyeroberts ready for review

vijaye12 committed:
change forecast to prediction

vijaye12 committed:
edit doc

vijaye12 committed:
doc edits

amyeroberts commented:
@kashif @ajati Holding off on a review until PatchTST has been merged. There's quite a few comments I would make e.g. removing the `PatchTSTTranspose` which have been [applied to PatchTST](https://github.com/huggingface/transformers/pull/25927#discussion_r1387015192) and will be reflected there after rebasing, running `make fix-copies` and updating the code.

namctin committed:
remove PatchTSMixerTranspose

namctin committed:
add PatchTSMixerPositionalEncoding and init position_enc

namctin committed:
remove positional_encoding

namctin committed:
edit forecast_masking, remove forecast_mask_ratios

ajati committed:
fix broken code

ajati committed:
var rename target_values -> future_values

ajati committed:
num_features -> d_model

ajati committed:
master merged

ajati committed:
fix broken code after master merge

ajati committed:
repo consistency

ajati committed:
use postional embedding

wgifford committed:
Merge remote-tracking branch 'upstream/main' into add-patchtsmixer

ajati committed:
prediction_logits -> prediction_outputs, make fix-copies

vijaye12 commented:
All the suggested changes from PatchTST, copied from are resolved in this branch too.. Ready for review and testcases passing now.

ajati committed:
uncommented @slow

vijaye12 committed:
minor changes

vijaye12 committed:
loss first in tuple

vijaye12 committed:
tuple and dict same ordering

vijaye12 committed:
style edits

vijaye12 committed:
minor changes

vijaye12 committed:
dict/tuple consistent enablement

vijaye12 commented:
@amyeroberts - Greetings! We have enabled all the changes suggested in PatchTST and also the corrections suggested in this PR from the past reviewers. Requesting your review and approval.

ArthurZucker reviewed (approved):
Thanks this looks a lot better than last time. Left some small nits but should be good to go otherwise

kashif committed:
Update src/transformers/models/patchtsmixer/modeling_patchtsmixer.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

kashif committed:
Update tests/models/patchtsmixer/test_modeling_patchtsmixer.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

kashif committed:
Update src/transformers/models/patchtsmixer/modeling_patchtsmixer.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

kashif committed:
fix formatting

kashif committed:
formatting

kashif committed:
usage tip

kashif committed:
test on cpu only

kashif committed:
Merge branch 'main' into add-patchtsmixer

vijaye12 committed:
add sample usage

vijaye12 committed:
change PatchTSMixerForClassification to PatchTSMixerForTimeSeriesClassification

vijaye12 committed:
push changes

vijaye12 committed:
fix copies

NielsRogge commented on docs/source/en/model_doc/patchtsmixer.md:
```diff
@@ -34,9 +34,26 @@ This model was contributed by [ajati](https://huggingface.co/ajati), [vijaye12](
 [wmgifford](https://huggingface.co/wmgifford), [kashif](https://huggingface.co/kashif).
 
+
+## Sample usage 
```
```suggestion
## Usage tips
```
It's called like this in the other docs

vijaye12 commented:
Hi @ArthurZucker -Thanks for approving this PR. We have resolved all the final changes you mentioned. Please review and help with merge, if all good.

PS: change with respect to adding segment names to docstring is the only one comment pending. When we add segment names in docstring following the syntax ssuggested - it getting auto removed during make fix-ups. Othan than this - all other. comments are resolved.

vijaye12 committed:
std scaling set to default True case

vijaye12 committed:
minor changes

vijaye12 committed:
stylechanges

kashif committed:
Merge branch 'main' into add-patchtsmixer

ArthurZucker commented:
@vijaye12 thanks for bearing with us in this long review! and congrats for the merge! 🚀
